src/D2L.CodeStyle.Analyzers/Language/LikelyArgumentMismatchAnalyzer.cs

```diff
@@ -0,0 +1,669 @@
+using System.Linq;
+using D2L.CodeStyle.Analyzers.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers.Language {
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	internal sealed class LikelyArgumentMismatchAnalyzer : DiagnosticAnalyzer {
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+			Diagnostics.LikelyArgumentMismatch
+		);
+
+		public override void Initialize( AnalysisContext context ) {
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction( RegisterAnalyzer );
+		}
+
+		private static void RegisterAnalyzer( CompilationStartAnalysisContext context ) {
+			context.RegisterSyntaxNodeAction(
+				ctx => AnalyzeInvocation(
+					ctx,
+					(InvocationExpressionSyntax)ctx.Node
+				),
+				SyntaxKind.InvocationExpression
+			);
+		}
+
+		private static void AnalyzeInvocation(
+			SyntaxNodeAnalysisContext context,
+			InvocationExpressionSyntax expr
+		) {
+			if( expr.ArgumentList == null ) {
+				return;
+			}
+
+			if( expr.ArgumentList.Arguments.Count <= 1 ) {
+				return;
+			}
+
+			// get bipartite matching of the source
+			ImmutableArray<Edge> sourceMatching = MatchArguments(
+				context.SemanticModel,
+				expr.ArgumentList.Arguments
+			).ToImmutableArray();
+
+			if( sourceMatching.Length <= 1 ) {
+				return;
+			}
+
+			ImmutableArray<Edge> bestMatching = GenerateBestMatching( context, sourceMatching );
+
+			ImmutableDictionary<ArgumentDetails, Edge> sourceEdges = sourceMatching
+				.ToImmutableDictionary( x => x.Arg, x => x );
+			foreach( Edge edge in bestMatching ) {
+				Edge sourceEdge = sourceEdges[ edge.Arg ];
+				if( !sourceEdge.Equals( edge ) ) {
+					context.ReportDiagnostic( Diagnostic.Create(
+						Diagnostics.LikelyArgumentMismatch,
+						sourceEdge.Arg.Location,
+						sourceEdge.Arg.Name,
+						edge.Param.Name,
+						sourceEdge.Param.Name
+					) );
+				}
+			}
+		}
+
+		private static ImmutableArray<Edge> GenerateBestMatching(
+			SyntaxNodeAnalysisContext context,
+			ImmutableArray<Edge> sourceMatching
+		) {
+			ImmutableArray<ArgumentDetails> arguments = sourceMatching.Select( x => x.Arg ).ToImmutableArray();
+			ImmutableArray<ParameterDetails> parameters = sourceMatching.Select( x => x.Param ).ToImmutableArray();
+
+			int[,] costMatrix = GetCostMatrix( context.SemanticModel, arguments, parameters );
+
+			int[] assignments = HungarianAlgorithm.FindAssignments( costMatrix );
+			ImmutableArray<Edge> matching = ConvertToMatching(
+				assignments,
+				arguments,
+				parameters
+			);
+
+			return matching;
+		}
+
+		private static int[,] GetCostMatrix(
+			SemanticModel model,
+			ImmutableArray<ArgumentDetails> arguments,
+			ImmutableArray<ParameterDetails> parameters
+		) {
+			int[,] matrix = new int[ arguments.Length, parameters.Length ];

```

Comment:
This needs to be padded with `int.MaxValue` to be a square matrix, otherwise the algo won't work

Reply:
I'm not sure why you'd believe me either!

Hm, rethinking this, I'm not sure we actually want to augment the matrix to be square, given what the cols/rows represent.  I think `params` and optional args are the cases we have to worry about?  Basically, how can we have a method invocation with more arguments than the function has parameters (`params`) or vice versa (  default values ).

For `params`, I think it would be unnecessary to suggest different orders.  Not really sure about optionals.

Reply:
Hmm, so I changed it back to only include args/params that were matched by non-excluded arguments. For instance, "trusting the dev" with a named parameter - should trust them for the arg _and_ the parameter.

Wasn't thinking about what this would look like for `params` or default values though... feels like defaulted parameters should definitely make their way in... perhaps similarly parameterss that had a literal passed.

So, somewhere in between what the two of us had would be good I think.